datajunction-server/datajunction_server/construction/build.py

```diff
@@ -281,6 +281,22 @@ def _build_tables_on_select(
         )  # got a materialization
         if node_table is None:  # no materialization - recurse to node first
             node_query = parse(cast(str, node.query))
+
+            dj_current_timestamp = list(
+                node_query.filter(
+                    lambda x: isinstance(x, ast.Function)
+                    and x.name.name.lower() == "dj_current_timestamp",
+                ),
+            )
+            if dj_current_timestamp and not node.has_available_materialization(
```

Comment:
Please remind me how would we render the query with the `${dj_current_timestamp}` variable (for scheduling purposes) if the build code blocks it with the check.

Reply:
That syntax doesn't look correct right? This is going to be a function executed in the DJ Runtime `DJ_CURRENT_TIMESTAMP()` which makes sense to me. is `${...}` Something else?

Reply:
@CircArgs and @agorajek So this is the part that's going to get a bit confusing and I was going to punt implementation to a separate PR 😄 

But the gist of it is this:
* At Netflix we wouldn't want DJ to actually generate any formatted timestamp strings, because we'll have a plugin in our internally deployed DJ server that just replaces the `DJ_CURRENT_TIMESTAMP()` function with something like `${dj_current_timestamp}`. The latter is a variable that gets substituted with the datetime parameters by our internal scheduler. 
* However, in open source DJ needs to actually execute the function to generate the right query for materialization. This means that for each partition that gets backfilled, DJ will need to execute `DJ_CURRENT_TIMESTAMP()` to get the right value for that partition and then kick off the materialization job.
* I'm a little bit concerned about getting the OSS implementation right, since I think the path would be clearest if we were actively building out an integration with a scheduler like Airflow.

Reply:
@shangyian just to show everyone a quick Airflow reference. Found this at https://docs.astronomer.io/learn/airflow-sql#example-2-execute-a-query-with-parameters:
```
opr_param_query = SnowflakeOperator(
    task_id="param_query",
    snowflake_conn_id="snowflake",
    sql="param-query.sql",
    params={"date":mydatevariable}
)
```
and the param-query.sql is:
```
SELECT *
FROM STATE_DATA
WHERE date = {{ params.date }}
```

So notice that we need to generate the SQL with a template that then will be understood by a given scheduler. But whatever the string replacement patter is we will need to generate that SQL... so maybe this code (your PR) is good, but when we generate SQL for materialization purpose we could pass in a special flag to not fail here.

Reply:
@agorajek Thanks for digging into this! It looks like actually all we need to do is replace `DJ_CURRENT_TIMESTAMP()` with `{{ params.date }}` for Airflow, since that SQL would just be written as a part of the transform's query. Also in looking at those docs, what do you think about naming this function `dj_logical date()` instead?

Reply:
+1 to using `DJ_LOGICAL_TIMESTAMP()` as the name.

> At Netflix we wouldn't want DJ to actually generate any formatted timestamp strings, because we'll have a plugin in our internally deployed DJ server that just replaces the DJ_CURRENT_TIMESTAMP() function with something like ${dj_current_timestamp}

Is this something that could reasonably be a global server setting? In other words, on any given deployment of DJ, either you want the function to resolve to a formatted date/time or you want it to resolve to some generic string like `${dj_current_timestamp}` or `{{ params.date }}`. If that's so, maybe we can have a global setting `TIMESTAMP_PARAM=${dj_current_timestamp}` and in the processing logic for `DJ_CURRENT_TIMESTAMP()` we do something along the lines of:
```py
if settings.timestamp_param:
    # replace DJ_CURRENT_TIMESTAMP() with settings.timestamp_param
else:
    # actually resolve to a formatted timestamp strings
```

Reply:
> Is this something that could reasonably be a global server setting?

Yes, and I think we may just hardcode few of these methods as options:
```
DJ_DEFAULT_TIMESTAMP_FORMAT: double_curly | dollar_with_curly | value_as_str | value_as_int
```
Remember that the above setting would only be used if the user does not specify format string as part of the DJ_LOGICAL_TIMESTAMP itself.

Reply:
Yeah, given that DJ is not doing the scheduling, I'm leaning towards:
* We have a server-side global config `DJ_DEFAULT_TIMESTAMP_FORMAT` that's one of the options @agorajek listed above.
* We don't bother implementing any sort of actual formatted timestamp (i.e., parsing a timestamp with the format string `%Y-%m-%d`), since if we integrate with an external scheduler (with the example above being Airflow), they will typically provide that generation for us and we just need to abide by their syntax.
* If DJ ever builds its own scheduler, then we should think about supporting a formatted timestamp.

Reply:
Can you help me understand why the global setting value has to be one of an "approved" set of variable formats and then having the user separately pass in the variable name? Wouldn't it be the same variable name that's always used across a single server, such as ${date_partition}?

Reply:
@samredai the global setting isn't about the variable name but rather how the variable substitution format should look like. Options include:
* `{{ date }}`
* `${date}`
* `$date`
etc

Reply:
I was actually wondering if the global setting should be about the variable which is why I asked that question. If the behavior desired is that DJ_LOGICAL_TIMESTAMP() always resolves to a parameter, why don't we just let the user directly define that parameter in the global settings, formatting included.

Reply:
Yep, that's what the current implementation looks like -- there's just a single global parameter that replaces DJ_LOGICAL_TIMESTAMP() across the board. That way we can be more constrained or else someone is going to use this to do some odd variable substitution that's not timestamp-based at all... (I just pushed the changes + updated the PR description on it)

Reply:
@shangyian I'm not saying this is some better alternative, but if we had the concept of variables would that be going too far? 

```sql
Node(
...
query= SELECT *
FROM STATE_DATA
WHERE date = my_date
...
variables = ['my_date']
```

at runtime

`GET dj.server/.../...&my_date='{{ date}}'`

so the user could fill in the variable knowing there is some downstream thing looking to dump into a templated part or they could specify an actual date in this case.

Reply:
The netflix specific runtime behavior of dumping a different format string as well as having this logic in the build for one case feels odd to me

Reply:
@CircArgs yeah I do think variables would be going a bit too far. Arbitrarily templated queries would create a huge amount of ambiguity across the DJ DAG

Reply:
@CircArgs So I'm not sure that's Netflix specific -- if we integrate in OSS with Airflow, we'll need something similar for incremental materialization of transforms. It's just that the substituted string would be different, and that's configurable via settings. But I agree that it's better if it's not in the build

Reply:
I didn't quite mean that that specific formatting method is Netflix specific, just questioning whether there should be a fixed set of options on how to format or if it should be something generic to my point about these being user-defined runtime functions

Reply:
> if we had the concept of variables would that be going too far?

@CircArgs heh, I'm afraid that would give too much power to the user. I think we should keep this functionality as locked down as possible (like right now it's basically limited to a single, temporal parameter), or at least until we have a clear use case. Otherwise we'll start going down the path of supporting jinja templates... 😄 

Also I removed the function-specific check from the build logic -- I realized that we don't need to stop the user from generating SQL for a node just because an upstream uses `dj_logical_timestamp()`. If their query engine doesn't support that, it'll just fail when they try to run the query, but we might as well let them generate it.